editor.lib/src/org/netbeans/editor/PopupManager.java

```diff
@@ -478,8 +478,10 @@ private static Placement determinePlacement(Placement placement, Dimension prefS
                 KeyStroke ks = KeyStroke.getKeyStrokeForEvent(e);
                 Object obj = im.get(ks);
                 LOG.log(Level.FINE, "Keystroke for event {0}: {1}; action-map-key={2}", new Object [] { e, ks, obj }); //NOI18N
-                if (obj != null && !obj.equals("tooltip-no-action") //NOI18N ignore ToolTipSupport installed actions
-                ) {
+                /* NETBEANS-403: Avoid forwarding actions into simple text popups, such as those
+                created by ToolTipSupport.setToolTipText. The actions will still work if the
+                popup receives explicit focus. */
+                if (obj != null && !(popup instanceof JTextComponent)) {
```

Comment:
Sorry for possibly silly questions; ToolTipSupport.setToolTipText seems to create either JEditorPane or JTextArea both of which descend from JTextComponent ?

Reply:
That's correct.

An alternative here would have been to define a new client property that can be set on any popup component (via JComponent.putClientProperty) that does not want actions forwarded to it, and have ToolTipSupport.setToolTipText set this property on the tooltip components that it creates.

But when I thought about this, I could think of no case where it makes sense to forward the event to the popup is the latter is a JTextComponent--the popup would just gobble up events that should really be handled by the editor, like it's doing currently. And editor actions like caret-end and caret-begin don't even work properly when invoked this way (they end up being applied on the editor instead, even if they came from the tooltip).

Arguably, action forwarding into popups should be opt-in rather than opt-out, but it seems that o.n.spi.debugger.ui.ToolTipUI depends on the current behavior. So I didn't try to undo the forwarding mechanism entirely.

Reply:
Hm, for example both code folding support and brace matching support actually add a JTextComponent to the popup (although indirectly, embedding into JPanel); in case of folding, the component might even benefit from handling the input. Also debugger may add a JTextComponent. 

I would rather advise to use the client property anyway, to explicitly control the tooltip's keyboard processing. The popup instance can be installed (I think) by a public API PopupManager.install(), and this change filter key events from such a component.

Also something should be done around ToolTipSupport.NO_ACTION usages; it uses "tooltip-no-action" constant whose handling is going to be changed: see  ToolTipSupport, line 392, 477, 1015-1027

Reply:
Is there ever a case for the popup to handle text editing events, though? If I type "Home" or "End" while the editor has the keyboard focus, I always expect the action to be carried out by the editor, not the popup. Besides, forwarded text editing events don't work properly in any case the way this code is written (probably because there is a null ActionEvent).

What about only forwarding events that are not handled by the editor itself?

Not sure what to do about NO_ACTION--it seems to have a different role, disabling popup actions whether or not the keyboard focus is in the popup.

If you still think a client option is needed--should it be opt-in, or opt-out wrt. keyboard forwarding?

Reply:
https://netbeans.org/bugzilla/show_bug.cgi?id=262311 seems to be the reason for introduction of the "tooltip-no-action".

Reply:
To maintain compatibility, the client option should be opt out - that is disabling the keyboard redirection. Existing clients may rely on the de-facto state.

Reply:
The action forwarding behavior has only been present since 8.2, since the forwarding logic was dormant due to the "!obj.equals("tooltip-no-action") && obj.equals("tooltip-hide-action")" bug before that (see JIRA issue). So it's unlikely that external plugins (outside the NetBeans codebase) have had time to depend on the action forwarding behavior.

My concern is that the "forward actions by default" approach will lead to more bugs like this in the future.

Reply:
OK, so if I understand finally correctly, then:
* there was a defect < 8.2, which caused the forward (and later the consume) code branch executed if and only if `tooltip-hide-action` was in the input map. I guess that ! was missing in there `(!obj.equals() && !obj.equals())`. This action worked despite it was forwarded to tooltip because it doesnt' use ActionEvent at all.
* in 8.2, it was corrected so that **all** actions except `tooltip-no-action` were forwarded from the editor to the popup, if the popup handled them. This is what causes `NETBEANS-403` and what you're trying to fix.

So the feature that tooltip component receives keystrokes for handled action was present since ~2005(?) till release 8.2 (when it was disabled by a bug). Now you want to remove it. Note that specifically mutation actions were changed to NO_ACTION, so the intention was clearly to have navigational actions forwarded to the tooltup. I don't have the CVS from pre-2007 days handy, but I suspect it might be related to e.g. accessibility.

Let's discuss more in the issue, because I can't reproduce the bug described in NETBEANS-403 at all in neither 8.2 (which I use for development) or a build from fresh sources.

Reply:
> the intention was clearly to have navigational actions forwarded to the tooltip
That seems like a usability problem. The actions of my keyboard should not depend on where on the screen my mouse pointer is. I know of no other piece of software that does this, except possibly X Windows ca. 1992. Tooltips are ephemeral things; users are unlikely to understand that the behavior of a particular keyboard action depends on whether a tooltip is showing or not. For myself, it took two weeks to discover the correlation--and I was hitting this bug dozens of times times per day and actively trying to find out what caused it. Even if this was the behavior between 2005 and 2010 (I didn't realize that), I suggest changing it.

It's still perfectly OK to enable _additional_ actions when a tooltip is showing, though.

Though if you hear these objections and still think the current behavior should be kept, I'm fine with that--the most important thing is that this particular bug is fixed.

=== Reproducing the bug ===

I am able to produce the bug on both NetBeans 8.2 and the latest master branch (acefa191423bf4...). The easiest is to open a long Java file, type in "asdfasdf" somewhere in the code, hover over the underlined error, wait for the popup to appear, and then press "End" on the keyboard (external keyboard if you don't have "End" on your laptop). See https://issues.apache.org/jira/secure/attachment/12911028/HomeEndBug.png . Does this not scroll to the bottom of the file for you. Are you on MacOS or Windows?

On MacBook keyboards, the "End" action can be accessed by Fn+RightArrow, but pressing the "Fn" button immediately hides the popup, so the problem doesn't occur in normal usage (for this particular action anyway). You can still trigger the error by holding down Fn before hovering to the error, then pressing RightArrow once the popup is showing.

Reply:
@eirikbakke no, I tested on Linux; I'll hope to reach some colleagues with Mac and Windows today.

Reply:
Update: verified on MacOS X; still I'd rather be very conservative, since some amount of other code is dedicated to forward whatever actions to bounce key events to the popup first (so it can override). But I still can't find working use-cases: the redirected action works on `null` event, so it will effectively work on the focused text component (which is the main editor in my case). Giving up.

Reply:
Just let me know your preferred course of action, and I can make a new patch.